### ISSUE TITLE
Form serialization does not include values set in web components

### ISSUE THREAD

**Issue by Tam** [jquery/jquery#5245]
### Description ###
Web components can set data on a form using [setFormValues](https://developer.mozilla.org/en-US/docs/Web/API/ElementInternals/setFormValue), but this isn't picked up by jQuery's form serialization.

### Link to test case ###
https://codepen.io/_Tam_/pen/yLRXYxe

**Comment by mgol** [comment]
Thanks for the report. jQuery traverses `form.elements`, reading values from each entry. Unfortunately, for a Web Component, reading the `value` property won't help.

Is there a way to read submitted form values from a Web Component? I guess no?

We may consider using `FormData` directly when available in `serializeArray` but we'll need to make sure all the current tests pass in the Web Component context as well.

If you'd like to take a stab at this, the code is here: https://github.com/jquery/jquery/blob/89ef81f86f8f371154e9fd3173be5fb57cb72d5e/src/serialize.js#L95-L97

**Comment by dmethvin** [comment]
I think the web component used in the example isn't fully implemented. If there were a `value` property getter/setter defined it should work. It's the job of the WC implementation to take whatever internal values there are and expose them to make it look like a native element.

**Comment by dmethvin** [comment]
More info:
* https://web.dev/more-capable-form-controls/
* https://javascript.plainenglish.io/web-components-can-now-be-native-form-elements-107c7a93386

Just tracing the code paths, it seems like `.val()` will get the value once the getter is defined.

**Comment by Tam** [comment]
I've updated the pen to include the additional getters / setters, but it hasn't had any effect on the output: https://codepen.io/_Tam_/pen/jOewRVq

I think the issue comes down to this part of the code filtering out any non-native form elements:
https://github.com/jquery/jquery/blob/89ef81f86f8f371154e9fd3173be5fb57cb72d5e/src/serialize.js#L104-L111

**Comment by dmethvin** [comment]
Yeah I agree, I didn't look far enough down the serialization to see the other roadblocks. I can't find  tickets but I'm pretty sure that `serializeArray` got those guards because people would select a group of elements, some of which weren't form elements, and be upset when they ended up in the serialization just because they had a `name` or `checked` property.

In this situation, though, we know for certain that the `elements` prop has the list we need, including the custom elements. The problem is that we do the filtering further down. I think a solution might involve skipping the extra filtering checks for a collection that came directly from `elements`.

**Comment by timmywil** [comment]
Dave to the rescue! We can look into avoiding filtering for `.elements`.

**Comment by claviska** [comment]
This affects [Web Awesome's](https://webawesome.com/) custom elements, which are form-associated. I created a small reproduction here, if it helps:

[View repro in CodePen](https://codepen.io/team/webawesome/pen/PwzaggN?editors=1010)

Note that form-associated custom elements _will_ appear in the `elements` collection, so perhaps allowing for tag names that contain a dash + checking for name/value would suffice here, e.g.:

```js
} ).filter( function() {
    var type = this.type;
    var nodeName = this.nodeName.toLowerCase();

    // Check if it's a custom element (contains a dash)
    var isCustomElement = nodeName.includes( "-" ) && "value" in this;

    // Use .is( ":disabled" ) so that fieldset[disabled] works
    return this.name && !jQuery( this ).is( ":disabled" ) &&
        ( rsubmittable.test( nodeName ) || isCustomElement ) && !rsubmitterTypes.test( type ) &&
        ( this.checked || !rcheckableType.test( type ) );
} ).map( function( _i, elem ) {
```

**Comment by TiesWestendorp** [comment]
Very annoying issue. Ran into it this just yesterday. Seconding the approach mentioned above. Alternatively, accepting a parameter that identifies the custom components somehow would also work.

**Comment by mgol** [comment]
PR at #5774. @Tam @claviska @TiesWestendorp could you overwrite `jQuery.fn.serializeArray` with an implementation from that PR locally and double check it solves your issues with not supporting form-associated Custom Elements?

**Comment by mgol** [comment]
BTW, I've been thinking about using `FormData` here. We'd get all that filtering we're doing out of the box. There are a few issues, though:
1. `valHooks` would be skipped. Now, outside of IE we now only have them for multi-select and `FormData` already handles multi-select identically as we do. Therefore, it's mostly about hooks defined by userland code, plugins, etc. I'm not sure how common are those, but we could re-evaluate for 5.0.
2. FormData works nicely on forms, but `.serialize()` can also be called on a collection of form elements, even if they're not lying in a form. I wish we didn't do this, serialization should always happen on the form level, but we do support it. I wonder how common such usage is compared to calling `.serialize()` on a form, but I wouldn't be surprised if it's common.

**Comment by mgol** [comment]
I extracted my FormData thoughts into https://github.com/jquery/jquery/issues/5778. Milestone set to `5.0.0`, but we're open to experimenting with it on the `4.x` line as long as we add some post-processing to apply `valHooks`.

**Comment by mgol** [comment]
@Tam @claviska @TiesWestendorp The PR was merged, so you can test with https://releases.jquery.com/git/jquery-git.js. Just don't deploy this version on a real public site, but some feedback that it works would be good to have.

**Comment by claviska** [comment]
@mgol thanks for the fix! I can confirm this works. Here's an updated repro with the latest:

https://codepen.io/team/webawesome/pen/PwGojME?editors=1010

Both controls are submitted as expected due to the custom element not being skipped.

**Comment by mgol** [comment]
I've noticed @Tam's test case at https://codepen.io/_Tam_/pen/jOewRVq still doesn't work and actually crashes now. This is because the `value` getter returns an array and it should be a string holding the first value even for multi-selects.

This may be more complex to make work with the current infrastructure. For multi-selects, jQuery iterates through options directly and reads values of selected ones. This may be hard to replicate with custom elements.

**Comment by mgol** [comment]
I've just had another look and that Codepen is actually broken. The custom element keeps `#value` internally as the whole FormData which of course extends beyond the form, causing the `value` getter to return a nested array.

Web Awesome single & multi-select works fine despite the multi-select returning an array in the `value` getter. I added an example with native single/multi select + Web Awesome single/multi-select and jQuery returns the same string as the FormData + URLSearchParams combination: https://codepen.io/mgol/pen/WbGvpGz?editors=1010
